### PR TITLE
Build QT version on OS X

### PR DESCRIPTION
**Pull request by mepard**
Changes to get WTF, WebCore, and JavaScriptCore to build for QT on OS
X. WebKit is not yet building.

### CODE REVIEW

**Comment by mepard** [comment]
Here are the commands I used:

```
brew install qt5
brew install jpeg
brew install PkgConfig

brew link --force qt5
ln -s /usr/local/Cellar/qt5/5.5.1_2/mkspecs /usr/local/mkspecs
ln -s /usr/local/Cellar/qt5/5.5.1_2/plugins /usr/local/plugins

cd ~/Projects/NewStuff/webkit-annulen-me/

Tools/Scripts/build-webkit  --debug --qt --cmakeargs="--trace  -DCMAKE_PREFIX_PATH=/usr/local/opt/qt5/" >buildoutQT.txt 2>&1
```

**Comment by mepard** [comment]
Here is the output from the build. Near the end you can see the errors building WebKit.

[buildoutQT.txt](https://github.com/annulen/webkit/files/136851/buildoutQT.txt)

**Comment by annulen** [comment]
Thanks for you contribution! See comments inline.
As for compilation errors, I've pushed patches that should fix them.

File: Source/JavaScriptCore/API/JSBase.h [hunk 1]
```diff
@@ -136,7 +136,7 @@ JS_EXPORT void JSGarbageCollect(JSContextRef ctx);
 
 /* Enable the Objective-C API for platforms with a modern runtime. */
 #if !defined(JSC_OBJC_API_ENABLED)
-#define JSC_OBJC_API_ENABLED (defined(__clang__) && defined(__APPLE__) && !defined(BUILDING_GTK__) && ((defined(__MAC_OS_X_VERSION_MIN_REQUIRED) && !defined(__i386__)) || (defined(TARGET_OS_IPHONE) && TARGET_OS_IPHONE)))
+#define JSC_OBJC_API_ENABLED (defined(__clang__) && defined(__APPLE__) && !defined(BUILDING_GTK__) && !defined(BUILDING_QT__) && ((defined(__MAC_OS_X_VERSION_MIN_REQUIRED) && !defined(__i386__)) || (defined(TARGET_OS_IPHONE) && TARGET_OS_IPHONE)))
```
Comment (annulen) [inline Source/JavaScriptCore/API/JSBase.h]: Couldn't we get away just defining JSC_OBJC_API_ENABLED=0, e.g. in JavaScriptCore/PlatfromQt.cmake?

File: Source/JavaScriptCore/runtime/DatePrototype.cpp [hunk 2]
```diff
@@ -125,7 +125,7 @@ namespace JSC {
 
 enum LocaleDateTimeFormat { LocaleDateAndTime, LocaleDate, LocaleTime };
  
-#if OS(DARWIN) && USE(CF)
+#if OS(DARWIN) && USE(CF) && !PLATFORM(QT)
```
Comment (annulen) [inline Source/JavaScriptCore/runtime/DatePrototype.cpp]: There is no !PLATFORM(QT) here in Qt branch. Why did you need it?

File: Source/WTF/icu/unicode/udat.h [hunk 3]
```diff
@@ -0,0 +1,997 @@
+/*
```
Comment (annulen) [inline Source/WTF/icu/unicode/udat.h]: I guess these ICU changes won;t be needed after we obtain b3a7bd025429ec8ad0474d4e4e7e5b16bf9304df. I'll try to do rebase tomorrow, but you can cherry-pick it locally for now.

File: Source/WTF/wtf/Platform.h [hunk 4]
```diff
@@ -581,10 +581,6 @@
 
 #endif /* PLATFORM(IOS) */
 
-#if PLATFORM(QT) && OS(DARWIN)
-#define USE_CF 1
```
Comment (annulen) [inline Source/WTF/wtf/Platform.h]: AFAIK USE_CF is required to use Mac implementation of MediaPlayer

File: Source/WTF/wtf/PlatformQt.cmake [hunk 5]
```diff
@@ -25,6 +25,23 @@ list(APPEND WTF_LIBRARIES
     ${CMAKE_THREAD_LIBS_INIT}
 )
 
+if (APPLE)
+   find_library(COCOA_LIBRARY Cocoa)
```
Comment (annulen) [inline Source/WTF/wtf/PlatformQt.cmake]: Please move all find_XXX() calls to OptionsQt.cmake. Also you may want to add REQUIRED option.

File: Source/WTF/wtf/WorkQueue.h [hunk 6]
```diff
@@ -34,7 +34,7 @@
 #include <wtf/RefCounted.h>
 #include <wtf/Threading.h>
 
-#if OS(DARWIN) && !PLATFORM(GTK)
+#if OS(DARWIN) && !PLATFORM(GTK) && !PLATFORM(QT)
```
Comment (annulen) [inline Source/WTF/wtf/WorkQueue.h]: Qt branch uses mac implementations of WorkQueue, see https://github.com/qtproject/qtwebkit/blob/dev/Source/WebKit2/Target.pri#L872

It was done to use Mac implementation of WebKit 2 IPC on OS X. I'm not really sure it will work for us now, but I guess it would be better to try it first. Also, look how WIN32 is implemented, it also uses Win sources.

File: Source/WebCore/PlatformQt.cmake [hunk 7]
```diff
@@ -218,7 +218,7 @@ if (USE_GSTREAMER)
     )
 endif ()
 
-if (ENABLE_VIDEO)
+if (ENABLE_VIDEO OR APPLE)
```
Comment (annulen) [inline Source/WebCore/PlatformQt.cmake]: Why? mediaControls stylesheet should not be used without VIDEO

File: Source/WebCore/xml/XSLTProcessorLibxslt.cpp [hunk 8]
```diff
@@ -50,7 +50,7 @@
 #include <wtf/text/StringBuffer.h>
 #include <wtf/unicode/UTF8.h>
 
-#if OS(DARWIN) && !PLATFORM(EFL) && !PLATFORM(GTK)
+#if OS(DARWIN) && !PLATFORM(EFL) && !PLATFORM(GTK) && !PLATFORM(QT)
```
Comment (annulen) [inline Source/WebCore/xml/XSLTProcessorLibxslt.cpp]: We should submit patch to upstream changing this condition to #if PLATFORM(COCOA)

File: Source/WebCore/xml/XSLTUnicodeSort.cpp [hunk 9]
```diff
@@ -35,11 +35,11 @@
 #include <libxslt/xsltutils.h>
 #include <wtf/unicode/Collator.h>
 
-#if OS(DARWIN) && !PLATFORM(EFL) && !PLATFORM(GTK)
+#if OS(DARWIN) && !PLATFORM(EFL) && !PLATFORM(GTK) && !PLATFORM(QT)
```
Comment (annulen) [inline Source/WebCore/xml/XSLTUnicodeSort.cpp]: Ditto.

File: Source/WebCore/xml/XSLTUnicodeSort.cpp [hunk 10]
```diff
@@ -35,11 +35,11 @@
...
 #include "SoftLinking.h"
 #endif
 
-#if OS(DARWIN) && !PLATFORM(EFL) && !PLATFORM(GTK)
+#if OS(DARWIN) && !PLATFORM(EFL) && !PLATFORM(GTK) && !PLATFORM(QT)
```
Comment (annulen) [inline Source/WebCore/xml/XSLTUnicodeSort.cpp]: Ditto.

File: Source/WebKit/mac/WebView/WebHTMLView.mm [hunk 11]
```diff
@@ -529,6 +529,8 @@ - (void)forwardContextMenuAction:(id)sender;
         return WebMenuItemTagMediaMute;
     case ContextMenuItemTagDictationAlternative:
         return WebMenuItemTagDictationAlternative;
+    case ContextMenuItemTagOpenLinkInThisWindow:
+        return Nullopt;		// FIXME: This was added to QTWebKit
```
Comment (annulen) [inline Source/WebKit/mac/WebView/WebHTMLView.mm]: Not needed anymore.

**Comment by annulen** [comment]
BTW, have you tried building using official Qt binaries?

File: Source/WebCore/xml/XSLTProcessorLibxslt.cpp (hunk 8 above)
Comment (annulen) [inline Source/WebCore/xml/XSLTProcessorLibxslt.cpp]: It turned out that there was PLATFORM(MAC) here originally, which was changed to this condition in https://bugs.webkit.org/show_bug.cgi?id=128464

BTW, what issue is causing soft linking for you? I guess if it works for Apple it should work for us as well.

I guess they use system-wide installed libxml2 and libxslt

**Comment by annulen** [comment]
@mepard: Would you mind to update your pull request? If all is well you should get working QtTestBrowser.

**Comment by mepard** [comment]
@annulen: Yeah, sorry I haven't been back here for a while. I'll give it a try over the weekend.

**Comment by annulen** [comment]
I've merged your changes to Source/JavaScriptCore/API/JSBase.h and Source/bmalloc/PlatformQt.cmake

We don't need to link Cocoa, CoreFoundation, and readline to WTF. Neither do we add include dirs for udis86 - they are already added in platform-independent part of build system

For ICU headers, could you try solution in this branch: https://github.com/annulen/webkit/tree/mac_icu - addition of new ICU headers should not be required anymore

**Comment by annulen** [comment]
All changes were integrated into qtwebkit-stable branch, build instructions are at https://github.com/annulen/webkit/wiki/Building-QtWebKit-on-OS-X.

@mepard: Please try to build it and report if there are any problems
